packages/cli/src/lib/vite/utils.ts

```diff
@@ -0,0 +1,73 @@
+import type {ServerResponse, IncomingMessage} from 'node:http';
+import path from 'node:path';
+import {Readable} from 'node:stream';
+import {Request, Response} from 'miniflare';
+import type {ViteDevServer} from 'vite';
+
+/**
+ * Creates a fully qualified URL from a Node request or a string.
+ * In the case of a Node request, it uses the host header to determine the origin.
+ */
+export function toURL(req: string | IncomingMessage = '/', origin?: string) {
+  const isRequest = typeof req !== 'string';
+  const pathname = (isRequest ? req.url : req) || '/';
+
+  return new URL(
+    pathname,
+    origin ||
+      (isRequest && req.headers.host && `http://${req.headers.host}`) ||
+      'http://example.com',
+  );
+}
+
+/**
+ * Turns a Node request into a Web request by using native Node APIs.
+ */
+export function toWeb(req: IncomingMessage, headers?: Record<string, string>) {
+  return new Request(toURL(req), {
+    method: req.method,
+    headers: {...headers, ...(req.headers as object)},
```

Comment:
Would this break with multiple of the same headers, like `Set-Cookie`?

Reply:
This one is fine because it's in the request object (there's no `set-cookie` here) but I'll double check the response. I had it in mind when I wrote the original code but I don't think I came back to fix it for `set-cookie` so great catch!